### ISSUE TITLE
FileSystemInfo

### ISSUE THREAD

**Issue by sergei66666** [dotnet/corefx#37743]
Hello.

I copy file via FTP from Windows to Linux. The file has Cyrillic characters. This leads to the fact that file names are displayed incorrectly.

And now the description of the problem.
I'm trying to get a list of all the files in a folder on Linux, and then delete them.

`
var filesToDelete = dir.EnumerateFileSystemInfos("*", SearchOption.AllDirectories).ToList();
FileSystemInfo file = filesToDelete.First();
`

At first, I used `File.Delete(file.FullName)` but without success. And this is logical, because the file name is incorrect.

Then I suggested that the call `file.Delete()` should work. But no.

I checked the `file.Attributes` property and it is -1.

But I expected to see the correct attributes.

My guess is the following.
The file name comes to `FileSystemInfo` as an array of bytes, then it is converted to a string as UTF8, and this string used to request all properties and `Delete` operation. But the string turns out to be incorrect due to the encoding mismatch and all operations failed.

Expected behavior (from my point of view):
Regardless of the encoding of the file name, we can request all attributes and delete the file if necessary.

ps. 
I use russian locale on windows 10.
ubuntu 16.04 with .net core 2.2.2 to run app.

**Comment by danmosemsft** [comment]
@JeremyKuhne

**Comment by joshudson** [comment]
I've been studying this problem for a long time. It is really really hard. I wish I could offer you something but I can't either.

**Comment by JeremyKuhne** [comment]
@sergei66666 can you provide some more detailed repro steps?

- What is a file name that does this?
- What ftp tool are you using?

Can you manipulate the file via the terminal? Or does it only present problems in .NET code?

**Comment by sergei66666** [comment]
1) On my windows 10 file name for example is "Тестовый файл.txt".
2) I transfer it on linux using WinSCP using FTP protocol. On linux installed vsftpd from repository (not patched version with support of filename conversion).
3) Via terminal I can type "ls -l" and see "-rw-r--r-- 1 vsftpd nogroup 0 May 28 04:52 ???????? ????.txt". I can't delete file, but I see attributes.

**Comment by joshudson** [comment]
His file system encoding isn't UTF-8 (only for some files). This borks the file system functions. This problem is hard to deal with for the same reason some tools upchuck on Windows when encountering a filename with unpaired surrogates.

**Comment by JeremyKuhne** [comment]
> His file system encoding isn't UTF-8

Yeah, this one is tricky. It could be potentially dealt with by including an extra reference in the FileSystemInfo class on Unix to hold the raw bytes if there is a conversion issue. I doubt we would want to take the space/performance hit for such a rare case.

One thing that might be worth doing is updating the recursive deletion to use the raw name bytes. It might allow removing stubborn folders and would be more performant as we could skip converting to UTF-16 and back.

**Comment by jhudsoncedaron** [comment]
I thought to fix it by making it possible to swap file system encodings at runtime using a [ThreadStatic] variable. Not really a good solution. Another possability is making a modified UTF-8 encoder-decoder pair that has no byte strings it can't round-trip. It's its own ugliness though.

**Comment by JeremyKuhne** [comment]
> I thought to fix it by making it possible to swap file system encodings 

Isn't the problem more of a one-off file with an encoding other than UTF-8? Or am I misreading?

**Comment by jhudsoncedaron** [comment]
> Isn't the problem more of a one-off file with an encoding other than UTF-8?

Probably. My home system has a mismash of UTF-8 and CP437 filenames since `/home` predates UTF-8 localization entries.

**Comment by JeremyKuhne** [comment]
I created #38028 to describe the improvement on deletion that I discussed [above](https://github.com/dotnet/corefx/issues/37743#issuecomment-496618562).
